File: Reqnroll/Chapter16/16.2-ContextShortcutPattern/Before/WIMP.Specs/Drivers/KitchenApiDriver.cs
using WIMP.App.Models;
using WIMP.Specs.Support;

namespace WIMP.Specs.Drivers;

public class KitchenApiDriver(RestApiContext restApiContext)
{
    public TestAction<Order> TakeNextOrder() =>
        new LambdaAction<Order>("Take next order", async () =>
            await restApiContext.ProcessRequest<Order>(
                "Take next order", HttpMethod.Post, "/api/kitchen/take-next-order"));

    public TestAction<VoidReturn> SetReady(int orderNo) =>
        new LambdaAction("Order ready to pick up", async () =>
            await restApiContext.ProcessRequest<VoidReturn>(
                "Order ready to pick up", HttpMethod.Post, $"/api/orders/{orderNo}/ready-for-pickup"));
}

File: CucumberJVM/Chapter19/19.2-SupplementaryAttachmentPattern/After/src/test/java/com/wimp/app/specs/drivers/KitchenApiDriver.java
package com.wimp.app.specs.drivers;

import com.wimp.app.models.Order;
import com.wimp.app.models.OrderCollectionDetails;
import com.wimp.app.specs.support.LambdaAction;
import com.wimp.app.specs.support.TestAction;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.service.annotation.HttpExchange;
import org.springframework.web.service.annotation.PostExchange;

@Component
public class KitchenApiDriver {
    private final KitchenApiClient kitchenApiClient;

    @HttpExchange("/api/kitchen")
    public interface KitchenApiClient {
        @PostExchange("/take-next-order")
        Order takeNextOrder();

        @PostExchange("/ready-for-pickup/{orderNo}")
        OrderCollectionDetails setReady(@PathVariable int orderNo);
    }

    @SuppressWarnings("SpringJavaInjectionPointsAutowiringInspection")
    public KitchenApiDriver(KitchenApiClient kitchenApiClient) {
        this.kitchenApiClient = kitchenApiClient;
    }

    public TestAction<Order> takeNextOrder() {
        return new LambdaAction<>("Take next order", kitchenApiClient::takeNextOrder);
    }

    public TestAction<OrderCollectionDetails> setReady(int orderNo) {
        return new LambdaAction<>("Set order as ready", orderNo, () -> kitchenApiClient.setReady(orderNo));
    }
}
